zkevm-circuits/src/evm_circuit/util/math_gadget/rlp.rs

```diff
@@ -0,0 +1,419 @@
+    param::{N_BYTES_ACCOUNT_ADDRESS, N_BYTES_U64, N_BYTES_WORD},
+    util::{constraint_builder::ConstraintBuilder, CachedRegion, Cell, RandomLinearCombination},
+};
+
+use super::IsZeroGadget;
+
+#[derive(Clone, Debug)]
+pub struct RlpU64Gadget<F> {
+    /// Byte representation of the U64 value.
+    value_rlc: RandomLinearCombination<F, N_BYTES_U64>,
+    /// Flag to mark the most significant byte in the U64's byte representation.
+    is_most_significant_byte: [Cell<F>; N_BYTES_U64],
+    /// Whether the most significant byte is zero, to check for zero value.
+    most_significant_byte_is_zero: IsZeroGadget<F>,
+    /// Boolean flag to mark whether or not the U64 value is less than 128.
+    is_lt_128: Cell<F>,
+}
+
+impl<F: Field> RlpU64Gadget<F> {
+    /// Configure and construct a gadget for RLP-encoding of a U64 value.
+    fn construct(cb: &mut ConstraintBuilder<F>) -> Self {
+        let value_rlc = cb.query_keccak_rlc();
+
+        let is_most_significant_byte = array_init::array_init(|_| cb.query_bool());
+        cb.require_boolean(
+            "at most one of is_most_significant_byte is one",
+            sum::expr(&is_most_significant_byte),
+        );
+
+        let most_significant_byte = sum::expr(
+            value_rlc
+                .cells
+                .iter()
+                .zip(&is_most_significant_byte)
+                .map(|(byte, indicator)| byte.expr() * indicator.expr()),
+        );
+        let most_significant_byte_is_zero = IsZeroGadget::construct(cb, most_significant_byte);
+
+        let value = expr_from_bytes(&value_rlc.cells);
+        cb.condition(most_significant_byte_is_zero.expr(), |cb| {
+            cb.require_zero("if most significant byte is 0, value is 0", value.clone());
+        });
+
+        for (i, is_most_significant) in is_most_significant_byte.iter().enumerate() {
+            cb.condition(is_most_significant.expr(), |cb| {
+                cb.require_equal(
+                    "most significant byte is non-zero",
+                    most_significant_byte_is_zero.expr(),
+                    0.expr(),
+                );
+                cb.require_equal(
+                    "higher bytes are 0",
+                    expr_from_bytes(&value_rlc.cells[0..(i + 1)]),
+                    value.clone(),
+                );
+            });
+        }
+
+        let is_lt_128 = cb.query_bool();
+        cb.condition(is_lt_128.expr(), |cb| {
+            cb.range_lookup(value, 128);
+        });
+
+        Self {
+            value_rlc,
+            is_most_significant_byte,
+            most_significant_byte_is_zero,
+            is_lt_128,
+        }
+    }
+
+    /// Assign witness data to the RlpU64 gadget.
+    fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        value: u64,
+    ) -> Result<(), Error> {
+        let value_bytes = value.to_le_bytes();
+
+        let most_significant_byte_index = value_bytes
```

Comment:
No, it would be different, since `position` will stop as soon as it finds a `true` (or a match for the closure). So whether we go in order or in reverse order matters.

In this case we have `value_bytes` in the Little-Endian byte format. So to find the most significant byte index, we need to reverse and then find the `position` at which `byte` in non-zero.